### PR TITLE
Add identifyUserV2 Action

### PR DESCRIPTION
**Pull request by mattgrosvenor-fs**
Note: This is into another local branch that will eventually be a PR into the main branch.

This PR adds a new action to the existing FullStory Cloud Mode destination. The new action recreates Identify using the [V2 endpoint](https://developer.fullstory.com/server/v2/users/create-user/).

## Testing

- [X] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [X] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

[ECO-6675]

[ECO-6675]: https://fullstory.atlassian.net/browse/ECO-6675?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      label: 'Traits',
+      default: {
+        '@path': '$.traits'
+      }
+    }
+  },
+  perform: (request, { payload, settings }) => {
+    const { properties, uid, email, display_name } = payload
+
+    const normalizedProperties = normalizePropertyNames(properties, { camelCase: true })
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:52]: Do we need to continue camel casing for backward compatibility?

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    delete normalizedProperties.email
+    delete normalizedProperties.name
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:55]: It feels like we should only delete `normalizedProperties.email` if `traits.email` is mapped to `email` and if `traits.name` is mapped to `display_name`. These are the default mappings, but a user can change these mappings for their destination action.

I don't think we have the ability to inspect mappings in this function. We could ensure `normalizedProperties.email === email` and `normalizedProperties.name === display_name` before deleting. However, that could lead to variable behavior if a customer sometimes provides matching values and other times different values depending on their `identify` payloads and destination mappings. I'm not sure I have a great solution for this.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    uid: {
+      type: 'string',
+      required: true,
+      description: "The user's id",
+      label: 'User ID',
+      default: {
+        '@path': '$.userId'
+      }
+    },
+    display_name: {
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:21]: Nit: Our other fields are camel-cased e.g. `displayName`. You could rename this field to `displayName` and still pass it to the API request body as `display_name`.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import type { Settings } from '../generated-types'
+import { createUserRequestParams } from '../request-params'
+import { normalizePropertyNames } from '../vars'
+
+const action: ActionDefinition<Settings> = {
+  title: 'Identify User V2',
+  description: 'Sets user identity variables, creates user if it does not exist',
+  platform: 'cloud',
+  defaultSubscription: 'type = "identify"',
+  fields: {
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:11]: Do we no longer provide a prescriptive mapping for `anonymousId`? In the v1 variant we map `anonymousId` to `segmentAnonymousId_str` in FullStory.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import { createUserRequestParams } from '../request-params'
+import { normalizePropertyNames } from '../vars'
+
+const action: ActionDefinition<Settings> = {
+  title: 'Identify User V2',
+  description: 'Sets user identity variables, creates user if it does not exist',
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:8]: Nit: This reads a little odd to me ("user"/"it"). Could go with something like:

"Sets user identity variables. Creates a FullStory user if a user matching the given uid doesn't already exist."

File: packages/destination-actions/src/destinations/fullstory/trackEventV2/generated-types.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+// Generated file. DO NOT MODIFY IT BY HAND.
+
+export interface Payload {
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/trackEventV2/generated-types.ts:3]: Is this file intended to be included as part of this PR? It looks like this PR only adds `identifyUserV2`.

File: packages/destination-actions/src/destinations/fullstory/index.ts [hunk 7]
```diff
@@ -21,6 +22,12 @@ const destination: DestinationDefinition<Settings> = {
       subscribe: 'type = "identify"',
       partnerAction: 'identifyUser',
       mapping: defaultValues(identifyUser.fields)
+    },
+    {
+      name: 'Identify User V2',
+      subscribe: 'type = "identify"',
+      partnerAction: 'identifyUserV2',
+      mapping: defaultValues(identifyUserV2.fields)
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/index.ts:30]: By including this as a preset, would a new FullStory cloud mode destination automatically get `identifyUserV2` configured? If so, I don't think we want that.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 4 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:11]: I wasn't sure on this one. I can keep anonymousId.

File: packages/destination-actions/src/destinations/fullstory/trackEventV2/generated-types.ts (hunk 6 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/trackEventV2/generated-types.ts:3]: must have accidentally grabbed that file while working on events.

File: packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts [hunk 8]
```diff
@@ -167,4 +167,39 @@ describe('FullStory', () => {
...
+
+      expect(response.status).toBe(200)
+      expect(JSON.parse(response.options.body as string)).toEqual({
+        uid: userId,
+        email,
+        display_name: displayName,
+        properties: {
+          originallyHyphenated: 'some string',
+          originallySpaced: true,
+          originallyDotted: 1.23
```
[the remark this answers is not in the document]
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts:200]: I believe it is retained. I'll add a test for it.

File: packages/destination-actions/src/destinations/fullstory/index.ts (hunk 7 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/index.ts:30]: I needed to add this to get the option to display in the test app. I don't know how to test your question.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 2 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:55]: Sounds like it is safer to leave the duplicates.

Comment (ndonze-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:55]: Agreed. Would have been nice to address the duplication, but I'm not sure we can do so in a safe way. Would be good to get Patrick's and Josh's thoughts on leaving the duplication in.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 4 above)
Comment (ndonze-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:11]: Josh may have some thoughts, but that would maintain closer field parity with our existing v1 action.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:52]: I don't know the history of camel casing. I don't believe it is required for V2.

Comment (ndonze-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:52]: If we're case sensitive on the server (I'm not sure if we are), I could see removing camel casing being a barrier for existing v1 action users migrating to v2. We also camel case in the browser variant, so I wonder if removing camel casing would lead to an unexpected divergence in behavior between our browser destination and cloud mode destination. I think as a general rule we want to be changing property names as little as possible, but in this case maybe there are good reasons to continue camel casing in the v2 variant. /cc @patrick-fs

File: packages/destination-actions/src/destinations/fullstory/__tests__/var.test.ts [hunk 9]
```diff
@@ -170,7 +170,7 @@ describe('normalizePropertyNames', () => {
       [expectedNamePreservingTypeSuffix]: obj[originalNameIncludingTypeSuffix]
     }
 
-    const actual = normalizePropertyNames(obj, { camelCase: true })
+    const actual = normalizePropertyNames(obj, { camelCase: true, typeSuffix: true })
     expect(actual).toEqual(expected)
   })
```
[the remark this answers is not in the document]
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/__tests__/var.test.ts:185]: Type suffixes are going away all together in V2 APIs and shouldn't be included in the V2 Segment action. If the V2 Segment Actions append type suffixes, then that would be a bug. Def good to verify that payloads aren't type suffixed.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:61]: > I think as a general rule we want to be changing property names as little as possible

Yes this is the goal, since we'd like to respect whatever naming conventions customers use for their custom data fields. @ndonze-fs I'm thinking that this original camel-case application is based off of V1 field name constraints where type suffixes needed to be a part of the field name. Otherwise, I'm not sure why we would force camel casing.

I'd feel a little sad about pulling forward an implementation that potentially changes our customer's data model (like: they're used to "not camel cased" but we force them to use camel case in FullStory).

Customers have to opt-in to the V2 destination. What if we made "we'll respect your naming conventions, but you might need to updates some FSTA queries" part of the opt-in agreement?

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 2 above)
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: Would another solution be to change the API contract such that `display_name` and `email` are reserved fields in the `properties` collection? (and not fields that live outside of properties)?

File: packages/destination-actions/src/destinations/fullstory/__tests__/var.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/__tests__/var.test.ts:185]: I'm kind of wondering if there shouldn't be a separate function all together: `normalizePropertyNamesV2` which simply doesn't append type suffixes (and possibly skips camel-casing - I have comments in a separate thread about this).

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (ndonze-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:61]: @patrick-fs The data points I'm thinking about:

- Our Segment browser destination [camel cases identify and setUserProperties custom vars](https://github.com/cowpaths/segmentio-action-destinations/blob/main/packages/browser-destinations/src/destinations/fullstory/identifyUser/index.ts#L68). I don't have the background for why this is. Maybe it was a convenient way at the time to strip invalid characters (e.g. spaces)?
- Our Segment browser destination [does not camel case event custom vars](https://github.com/cowpaths/segmentio-action-destinations/blob/main/packages/browser-destinations/src/destinations/fullstory/trackEvent/index.ts#L33).
- Since we were camel casing custom user vars in our Segment browser destination, we took the same approach with our [v1 identifyUser cloud mode destination](https://github.com/cowpaths/segmentio-action-destinations/blob/main/packages/destination-actions/src/destinations/fullstory/identifyUser/index.ts#L61).
- Since we were not camel casing custom event vars in our Segment browser destination, we took the [same approach with our v1 trackEvent cloud mode destination](https://github.com/cowpaths/segmentio-action-destinations/blob/main/packages/destination-actions/src/destinations/fullstory/trackEvent/index.ts#L73) -- we didn't type suffix custom event vars.

That seems to raise questions around whether breaking parity between our v1 browser destination and v2 cloud mode destination will be problematic and whether dropping camel casing on the v2 cloud mode destination will have ramifications for existing customers. To your point, we could communicate any such ramifications to customers making the migration.

Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:61]: > breaking parity between our v1 browser destination and v2 cloud mode destination will be problematic and whether dropping camel casing on the v2 cloud mode destination will have ramifications for existing customers

The implication is that the same set of user properties sent to browser destinations will be named differently when sent to server destinations, at least until there's a v2 browser destination.

I talked with Josh Rynne about not having a V2 browser destination available when the V2 cloud destination is available. He wasn't concerned at the time since he didn't think there was a lot of overlap between browser and cloud destination customers.

So, by that logic we might be all right. WDYT?

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 2 above)
Comment (ndonze-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: @patrick-fs I'm not sure. Imagine the following scenario:

A customer has applied the default mappings, meaning `traits.name` would be mapped to `display_name`. We can pass `properties.display_name` in the request body with the value given to `traits.name`. However, I don't see a way for us to know that `traits.name` was mapped to `properties.display_name` when the action is executing other than by comparing their values and taking a guess that they probably don't need a `name_str` custom var in FullStory. The fact that `display_name` is nested within `properties` wouldn't seem to help us.

We could change our default mappings such that we'd expect a customer to send `traits.display_name` to Segment instead of `traits.name`. However:
- Customers can change their mappings such that we could end up back in the first scenario where a customer could map `traits.name` to `properties.display_name`. We could potentially argue that the duplication is more warranted in that case because they've diverged from the default mappings.
- `traits.name` may have been purposefully chosen based on common Segment usage patterns. I don't have background here.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (ndonze-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:61]: If we're okay with breaking parity between v1 browser identify and v2 cloud mode identify -- and if we're comfortable with property names potentially changing between v1 cloud mode identify to v2 cloud mode identify (i.e. no longer being camel cased) -- I think we'd be all right. You mentioned us not being case sensitive on the server, so the casing alone wouldn't likely be an issue. If the camel casing lodash function was also changing other characters we could end up with legitimately different property names. We could dig into this more.

Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:61]: Looking at the lodash documentation for [camelCase](https://lodash.com/docs/#camelCase) and our own transformation, `stripCharsNotSupportedInPropertyNames`, we would be losing removal of `-` characters.

Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: >  if we're comfortable with property names potentially changing between v1 cloud mode identify to v2 cloud mode identify (i.e. no longer being camel cased)

I'm comfortable with this, since I think this actually gets us more to where we need to be in respecting our customers' data models. From this perspective: v1 mishandled their property names, and v2 corrects this.

> `stripCharsNotSupportedInPropertyNames`, we would be losing removal of `-` characters

In this case I suppose there are reasons why we can't support hyphens related to indexing? It would be a good idea to dig into why there are characters we can't support that are reasonable for use in field names. For now, leave it in.

File: packages/destination-actions/src/destinations/fullstory/request-params.ts [hunk 10]
```diff
@@ -139,3 +139,23 @@ export const deleteUserRequestParams = (settings: Settings, userId: string): Req
...
+
+/**
+ * Returns {@link RequestParams} for the set user properties HTTP API endpoint.
+ *
+ * @param settings Settings configured for the cloud mode destination.
+ * @param requestBody The request body containing user properties to set.
+ */
+
+export const createUserRequestParams = (settings: Settings, requestBody: Object): RequestParams => {
+  const defaultParams = defaultRequestParams(settings, 'v2beta/users')
```
[the remark this answers is not in the document]
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/request-params.ts]: Much conversation has been happening lately about supporting both `v2beat` and `v2` routes concurrently. It looks like y'all creaeted a ticket for it yesterday during backlog refinement: https://fullstory.atlassian.net/browse/ECO-8516

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: > For now, leave it in.
For clarity, are you saying to leave in the `camelCase` transformation?  I could modify the `stripCharsNotSupportedInPropertyNames` to strip `-`

Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: Sorry that was ambiguous:

> are you saying to leave in the camelCase transformation?

No sir, take that out.

> I could modify the stripCharsNotSupportedInPropertyNames to strip -

I meant don't make any changes outside of removing camelCase. The way I read the thread was that `-` is currently being removed, so leave that piece alone. But, it would be good to understand why we're stripping out characters that are otherwise safe for JSON field names.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 4 above)
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts:11]: Please be sure to not add type suffixes though. `_str` gets treated as part of the field name in the V2 API.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 2 above)
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: All right, I got it. So do I have this right:

* The two `delete` statements safe as long as default mappings are preserved
  * `delete normalizedProperties.email`
  * `delete normalizedProperties.email`
* But default mappings can be changed in the Segment UI and we have no way of knowing whether they've changed or what their state is at action execution time
* Therefor, we need to maintain duplicate data (`$. email`, `$.properties.email` + `$.display_name`, `$.properties.name`) to prevent losing data all together

Do I have that right?

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: I believe this is once again an attempt to keep names in line with the browser requirements as found [here](https://help.fullstory.com/hc/en-us/articles/360020623234#property-name-requirements).

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 2 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: Yes, you have that right.

File: packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts (hunk 1 above)
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/identifyUserV2/index.ts]: Cool, I was more getting to why we couldn't support things outside of alpha-numerics and `_`. No one on Eco can answer this question, I think the answer would come from the Indexing team.

In any case, we're good to move forward with removing the camel-case enforcement but keeping the character constraints.

File: packages/destination-actions/src/destinations/fullstory/request-params.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/request-params.ts]: For now I believe the plan is to merge this with `v2beta`.

File: packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts [hunk 11]
```diff
@@ -167,4 +167,45 @@ describe('FullStory', () => {
...
+        email,
+        display_name: displayName,
+        properties: {
+          email,
+          name: displayName,
+          segmentAnonymousId: anonymousId,
+          originallyhyphenated: 'some string',
+          originallyspaced: true,
+          originallydotted: 1.23,
+          typeSuffixed_bool: true
```
Comment (ndonze-fs) [inline packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts:206]: This seems like the correct behavior (not to strip a type suffix if one is given). To be sure, the v2 API will gracefully handle given type suffixes even though they're no longer required, correct?

**Review by ndonze-fs — APPROVED** [review]
Left a couple final questions, but this LGTM.

File: packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts (hunk 11 above)
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts:206]: > Please be sure to not add type suffixes though. _str gets treated as part of the field name in the V2 API.

Based on @patrick-fs comment above type suffixes will be ignored and treated as part of the field name.

File: packages/destination-actions/src/destinations/fullstory/request-params.ts [hunk 12]
```diff
@@ -139,3 +139,23 @@ export const deleteUserRequestParams = (settings: Settings, userId: string): Req
...
+
+/**
+ * Returns {@link RequestParams} for the V2 Create User HTTP API endpoint.
+ *
+ * @param settings Settings configured for the cloud mode destination.
+ * @param requestBody The request body containing user properties to set.
+ */
+
+export const createUserRequestParams = (settings: Settings, requestBody: Object): RequestParams => {
+  const defaultParams = defaultRequestParams(settings, `v2beta/users?${integrationSourceQueryParam}`)
```
[the remark this answers is not in the document]
Comment (mattgrosvenor-fs) [reply packages/destination-actions/src/destinations/fullstory/request-params.ts:151]: You are correct. @sabrina-li is moving integration source to request headers. Since extra headers would just be ignored. I can add this new header and it will be picked up when ready.

[the remark this answers is not in the document]
Comment (sabrina-li) [reply packages/destination-actions/src/destinations/fullstory/request-params.ts:151]: segment is special in that, if the `integration` is not specified (either in request or in header), the events api always attempt to check for `User-Agent` as well. With v2 we plan on removing the query param completely before GA so only header will be supported soon.
Also, I don't think the users endpoint has integration source (yet): there's no where to put it since it doesn't result in event bundles like the events api.
Anyways, I vote for putting integrations source to headers or both(req&header), it shouldn't hurt.

File: packages/destination-actions/src/destinations/fullstory/__tests__/var.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrick-fs) [reply packages/destination-actions/src/destinations/fullstory/__tests__/var.test.ts:185]: @ndonze-fs yeah my comment about `normalizePropertyNamesV2` was more about playing defense and mitigating the likelihood that our future selves or others accidentally flip a bit like `typeSuffix: true` which has a completely different impact on the downstream behavior of FSTA.
